File: backend-go/internal/util/permission_presets.go
package util

import "strings"

// ExpandPermissionPresets expands Java/frontend permission preset tokens (editor, full_crud, ...).
func ExpandPermissionPresets(permissions []string) []string {
	if len(permissions) == 0 {
		return permissions
	}
	out := append([]string{}, permissions...)
	for _, raw := range permissions {
		switch strings.ToLower(strings.TrimSpace(raw)) {
		case "viewer":
			out = MergeUniqueCaseInsensitive(out, []string{"view"})
		case "editor":
			out = MergeUniqueCaseInsensitive(out, []string{"view", "create", "edit"})
		case "full_crud":
			out = MergeUniqueCaseInsensitive(out, []string{"view", "create", "edit", "delete"})
		case "full_crud_export":
			out = MergeUniqueCaseInsensitive(out, []string{"view", "create", "edit", "delete", "export"})
		case "admin_full":
			out = MergeUniqueCaseInsensitive(out, []string{"admin", "view", "create", "edit", "delete", "export", "scope:all"})
		}
	}
	return out
}

File: backend-rust/src/util/permission_presets.rs
use super::permission_bitfield::PermissionBitfieldUtil;

/// Expand Java/frontend permission preset tokens (editor, full_crud, ...).
pub fn expand_permission_presets(permissions: &[String]) -> Vec<String> {
    if permissions.is_empty() {
        return Vec::new();
    }
    let mut out = permissions.to_vec();
    for raw in permissions {
        match raw.trim().to_ascii_lowercase().as_str() {
            "viewer" => {
                out = PermissionBitfieldUtil::merge_unique_case_insensitive(&out, &["view".into()]);
            }
            "editor" => {
                out = PermissionBitfieldUtil::merge_unique_case_insensitive(
                    &out,
                    &["view".into(), "create".into(), "edit".into()],
                );
            }
            "full_crud" => {
                out = PermissionBitfieldUtil::merge_unique_case_insensitive(
                    &out,
                    &[
                        "view".into(),
                        "create".into(),
                        "edit".into(),
                        "delete".into(),
                    ],
                );
            }
            "full_crud_export" => {
                out = PermissionBitfieldUtil::merge_unique_case_insensitive(
                    &out,
                    &[
                        "view".into(),
                        "create".into(),
                        "edit".into(),
                        "delete".into(),
                        "export".into(),
                    ],
                );
            }
            "admin_full" => {
                out = PermissionBitfieldUtil::merge_unique_case_insensitive(
                    &out,
                    &[
                        "admin".into(),
                        "view".into(),
                        "create".into(),
                        "edit".into(),
                        "delete".into(),
                        "export".into(),
                        "scope:all".into(),
                    ],
                );
            }
            _ => {}
        }
    }
    out
}
